## Biostat/Biomath M257 Homework 7
#### Due June 16 @ 11:59PM
### author: Sisi Shao

System information (for reproducibility):

In [1]:
versioninfo()

Julia Version 1.8.5
Commit 17cfb8e65ea (2023-01-08 06:45 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.5.0)
  CPU: 10 × Apple M1 Max
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores


Load packages:

In [2]:
using Pkg

Pkg.activate(pwd())
Pkg.instantiate()
Pkg.status()

  Activating project at `~/Downloads/SSS/Courses/Bio257_23Spring/HW/hw7`
    Updating registry at `~/.julia/registries/General.toml`
   Installed Parsers ───────────────── v2.7.0
   Installed HypergeometricFunctions ─ v0.3.17
   Installed FillArrays ────────────── v1.2.0
    Updating `~/Downloads/SSS/Courses/Bio257_23Spring/HW/hw7/Project.toml`
  [6e4b80f9] + BenchmarkTools v1.3.2
  [31c24e10] + Distributions v0.25.95
    Updating `~/Downloads/SSS/Courses/Bio257_23Spring/HW/hw7/Manifest.toml`
  [6e4b80f9] + BenchmarkTools v1.3.2
  [49dc2e85] + Calculus v0.5.1
  [d360d2e6] + ChainRulesCore v1.16.0
  [9e997f8a] + ChangesOfVariables v0.1.7
  [34da2185] + Compat v4.6.1
  [9a962f9c] + DataAPI v1.15.0
  [864edb3b] + DataStructures v0.18.13
  [b429d917] + DensityInterface v0.4.0
  [31c24e10] + Distributions v0.25.95
  [ffbed154] + DocStringExtensions v0.9.3
  [fa6b7ba4] + DualNumbers v0.6.8
  [1a297f60] + FillArrays v1.2.0
  [34004b35] + HypergeometricFunctions v0.3.17
  [3587e190] + InverseF

Status `~/Downloads/SSS/Courses/Bio257_23Spring/HW/hw7/Project.toml`
  [6e4b80f9] BenchmarkTools v1.3.2
  [31c24e10] Distributions v0.25.95
  [37e2e46d] LinearAlgebra
  [9a3f8284] Random


Again we continue with the linear mixed effects model (LMM)
$$
    \mathbf{Y}_i = \mathbf{X}_i \boldsymbol{\beta} + \mathbf{Z}_i \boldsymbol{\gamma}_i + \boldsymbol{\epsilon}_i, \quad i=1,\ldots,n,
$$
where   
- $\mathbf{Y}_i \in \mathbb{R}^{n_i}$ is the response vector of $i$-th individual,  
- $\mathbf{X}_i \in \mathbb{R}^{n_i \times p}$ is the fixed effects predictor matrix of $i$-th individual,  
- $\mathbf{Z}_i \in \mathbb{R}^{n_i \times q}$ is the random effects predictor matrix of $i$-th individual,  
- $\boldsymbol{\epsilon}_i \in \mathbb{R}^{n_i}$ are multivariate normal $N(\mathbf{0}_{n_i},\sigma^2 \mathbf{I}_{n_i})$,  
- $\boldsymbol{\beta} \in \mathbb{R}^p$ are fixed effects, and  
- $\boldsymbol{\gamma}_i \in \mathbb{R}^q$ are random effects assumed to be $N(\mathbf{0}_q, \boldsymbol{\Sigma}_{q \times q}$) independent of $\boldsymbol{\epsilon}_i$.

The log-likelihood of the $i$-th datum $(\mathbf{y}_i, \mathbf{X}_i, \mathbf{Z}_i)$ is 
$$
    \ell_i(\boldsymbol{\beta}, \mathbf{L}, \sigma_0^2) = - \frac{n_i}{2} \log (2\pi) - \frac{1}{2} \log \det \boldsymbol{\Omega}_i - \frac{1}{2} (\mathbf{y} - \mathbf{X}_i \boldsymbol{\beta})^T \boldsymbol{\Omega}_i^{-1} (\mathbf{y} - \mathbf{X}_i \boldsymbol{\beta}),
$$
where
$$
    \boldsymbol{\Omega}_i = \sigma^2 \mathbf{I}_{n_i} + \mathbf{Z}_i \boldsymbol{\Sigma} \mathbf{Z}_i^T.
$$
Given $m$ independent data points $(\mathbf{y}_i, \mathbf{X}_i, \mathbf{Z}_i)$, $i=1,\ldots,m$, we seek the maximum likelihood estimate (MLE) by maximizing the log-likelihood
$$
\ell(\boldsymbol{\beta}, \boldsymbol{\Sigma}, \sigma_0^2) = \sum_{i=1}^m \ell_i(\boldsymbol{\beta}, \boldsymbol{\Sigma}, \sigma_0^2).
$$

In HW6, we used the nonlinear programming (NLP) approach (Newton type algorithms) for optimization. In this assignment, we derive and implement an expectation-maximization (EM) algorithm for the same problem.

In [3]:
# load necessary packages; make sure install them first
using BenchmarkTools, Distributions, LinearAlgebra, Random

[ Info: Precompiling Distributions [31c24e10-a181-5473-b8eb-7969acd0382f]


## Q1. (10 pts) Refresher on normal-normal model

Assume the conditional distribution
$$
\mathbf{y} \mid \boldsymbol{\gamma} \sim N(\mathbf{X} \boldsymbol{\beta} + \mathbf{Z} \boldsymbol{\gamma}, \sigma^2 \mathbf{I}_n)
$$
and the prior distribution
$$
\boldsymbol{\gamma} \sim N(\mathbf{0}_q, \boldsymbol{\Sigma}).
$$
By the Bayes theorem, the posterior distribution is
\begin{eqnarray*}
f(\boldsymbol{\gamma} \mid \mathbf{y}) &=& \frac{f(\mathbf{y} \mid \boldsymbol{\gamma}) \times f(\boldsymbol{\gamma})}{f(\mathbf{y})}, \end{eqnarray*}
where $f$ denotes corresponding density. 

Show that the posterior distribution of random effects $\boldsymbol{\gamma}$ is a multivariate normal with mean
\begin{eqnarray*}
\mathbb{E} (\boldsymbol{\gamma} \mid \mathbf{y}) &=& \sigma^{-2} (\sigma^{-2} \mathbf{Z}^T \mathbf{Z} + \boldsymbol{\Sigma}^{-1})^{-1 } \mathbf{Z}^T (\mathbf{y} - \mathbf{X} \boldsymbol{\beta}) \\
&=& \boldsymbol{\Sigma} \mathbf{Z}^T (\mathbf{Z} \boldsymbol{\Sigma} \mathbf{Z}^T + \sigma^2 \mathbf{I})^{-1} (\mathbf{y} - \mathbf{X} \boldsymbol{\beta})
\end{eqnarray*}
and covariance
\begin{eqnarray*}
\text{Var} (\boldsymbol{\gamma} \mid \mathbf{y}) &=& (\sigma^{-2} \mathbf{Z}^T \mathbf{Z} + \boldsymbol{\Sigma}^{-1})^{-1} \\
&=& \boldsymbol{\Sigma} - \boldsymbol{\Sigma} \mathbf{Z}^T (\mathbf{Z} \boldsymbol{\Sigma} \mathbf{Z}^T + \sigma^2 \mathbf{I})^{-1} \mathbf{Z} \boldsymbol{\Sigma}.
\end{eqnarray*}

## Sol Q1

##### Method: conditional multivariate normal distribution

Define $W=\mathbf{y}-\mathbf{X\beta}$, then re-write the above as an augmented linear model:
\begin{align}
W &=\mathbf{Z}e_\gamma +e_W\\
\gamma&=e_\gamma\\
e_W&\sim\mathcal{N}(0,\sigma^2I)\\
e_\gamma&\sim\mathcal{N}(0,\Sigma)
\end{align}

Then immediately we have
$$\left[\begin{matrix}W\\\gamma\end{matrix}\right]\sim\mathcal{N}\left(\left[\begin{matrix} 0\\0\end{matrix}\right],\left[\begin{matrix}Z\Sigma Z^T+\sigma^2I& Z\Sigma\\\Sigma Z^T &\Sigma \end{matrix}\right]\right)$$

Thus, by conditional multivariate normal formula in *Biostat 200C* and *Biostat 250A*, we have

$$\gamma|W\sim\mathcal{N}(\mu_{\gamma|W},\Sigma_{\gamma|W})$$

where
\begin{align}
\mu_{\gamma|W}&=\Sigma Z^T\left(Z\Sigma Z^T+\sigma^2I\right)^{-1}W\\
\Sigma_{\gamma|W}&=\Sigma-\Sigma Z^T(Z\Sigma Z^T+\sigma^2I)^{-1}Z\Sigma
\end{align}

Subsituting $W$ with $y-X\beta$ will reproduce Dr. Zhou's solution.

## Q2. (20 pts) Derive EM algorithm

1. Write down the complete log-likelihood
$$
\sum_{i=1}^m \log f(\mathbf{y}_i, \boldsymbol{\gamma}_i \mid \boldsymbol{\beta}, \boldsymbol{\Sigma}, \sigma^2)
$$

2. Derive the $Q$ function (E-step).
$$
Q(\boldsymbol{\beta}, \boldsymbol{\Sigma}, \sigma^2 \mid \boldsymbol{\beta}^{(t)}, \boldsymbol{\Sigma}^{(t)}, \sigma^{2(t)}).
$$

3. Derive the EM (or ECM) update of $\boldsymbol{\beta}$, $\boldsymbol{\Sigma}$, and $\sigma^2$ (M-step). 

## Sol Q2

## Sol Q2

#### Q2.1

- By Q1, the complete likelihood for a single datum is indeed
\begin{align}
\exp\left(\mathcal{L}_i(\beta,\Sigma,\sigma^2)\right)&=\frac{1}{(2\pi)^{n_i/2}\sigma^{n_i}}\exp\{-\frac{1}{2\sigma^2}\lVert y_i-X_i\beta-Z_i\gamma_i \lVert_2^2\}\times\frac{1}{(2\pi)^{q/2}|\Sigma|^{1/2}}\exp\{-\frac{1}{2}\gamma_i^T\Sigma^{-1}\gamma_i\}\\
&=\frac{1}{(2\pi)^{\frac{q+n_i}{2}}(\sigma^2)^{\frac{n_i}{2}}|\Sigma|^{\frac{1}{2}}}\exp\left\{-\frac{1}{2\sigma^2}\left(\lVert y_i-X_i\beta-Z_i\gamma_i \lVert_2^2\ +\ \sigma^2\gamma_i^T\Sigma^{-1}\gamma_i\right)\right\}
\end{align}

Hence, the log-likelihood is

$$\mathcal{L}_i(\beta,\Sigma,\sigma^2)=-\left[\frac{q+n_i}{2}\log2\pi+\frac{n_i}{2}\log\sigma^2+\frac{1}{2}\log|\Sigma|\right]-\frac{1}{2\sigma^2}\left[\lVert y_i-X_i\beta-Z_i\gamma_i\lVert_2^2+\sigma^2\gamma_i^T\Sigma^{-1}\gamma_i\right]$$

Therefore, the complete log-likelihood for the whole dataset is 
\begin{align}
\mathcal{L}(\beta,\Sigma,\sigma^2)&=\sum_{i=1}^m\mathcal{L}_i(\beta,\Sigma,\sigma^2)\\
&=-\left[\frac{mq+n}{2}\log2\pi+\frac{n}{2}\log\sigma^2+\frac{m}{2}\log|\Sigma|\right] - \frac{1}{2\sigma^2}\left[\sum_{i=1}^m\lVert y_i-X_i\beta-Z_i\gamma_i\lVert_2^2+\sigma^2\sum_{i=1}^m\gamma_i^T\Sigma^{-1}\gamma_i\right]
\end{align}

#### Q2.2
First, note that
$$\gamma_i^TZ_i^TZ_i\gamma_i=\text{Tr}(\gamma_i\gamma_i^TZ_i^TZ_i)$$
$$\gamma_i^T\Sigma^{-1}\gamma_i=\text{Tr}(\gamma_i\gamma_i^T\Sigma^{-1})$$

Let $w_i=y_i-X_i\beta^{(t)}$, then by Q1, we know that the first and second **geometric moments** are

\begin{align}
\mu_{1}^{i}=\mathbb{E}\left(\gamma_i\mid \boldsymbol{\beta}^{(t)}, \boldsymbol{\Sigma}^{(t)}, \sigma^{2(t)}\right)&=\sigma^{-2(t)}\left(\frac{Z_i^TZ_i}{\sigma^{2(t)}}+(\boldsymbol{\Sigma}^{(t)})^{-1}\right)^{-1}Z_i^Tw_i\\
\mu_{2}^{i}=\mathbb{E}\left(\gamma_i\gamma_i^T\mid \boldsymbol{\beta}^{(t)}, \boldsymbol{\Sigma}^{(t)}, \sigma^{2(t)}\right)&=\left(\frac{Z_i^TZ_i}{\sigma^{2(t)}}+(\boldsymbol{\Sigma}^{(t)})^{-1}\right)^{-1}+\sigma^{-4(t)}\left(\frac{Z_i^TZ_i}{\sigma^{2(t)}}+(\boldsymbol{\Sigma}^{(t)})^{-1}\right)^{-1}Z_i^Tw_iw_i^TZ_i\left(\frac{Z_i^TZ_i}{\sigma^{2(t)}}+(\boldsymbol{\Sigma}^{(t)})^{-1}\right)^{-1}
\end{align}

Hence, the $\mathcal{Q}$ function is

$$\mathcal{Q}=C-\frac{n}{2}\log\sigma^2-\frac{m}{2}\log|\Sigma|+\frac{1}{2\sigma^2}\left[\sum_{i=1}^m\lVert w_i\lVert_2^2+\sigma^2\sum_{i=1}^m\text{Tr}\left(\mu_2^i \left(\frac{Z_i^TZ_i}{\sigma^2}+\Sigma^{-1}\right)\right)-2\sum_{i=1}^mw_i^TZ_i\mu_1^i\right]$$

where $C$ is a constant independent of parameters.

#### Q2.3

Take derivatives w.r.t. all parameters and set them to 0,

\begin{align}
\widehat{\sigma^2}^{(t+1)}&=\frac{1}{n}\sum_{i=1}^m\left[\lVert\widehat{w}_i\lVert_2^2+\text{Tr}(\mu_2^iZ_i^TZ_i)-2w_i^TZ_i\mu_1^i\right]\\
\widehat{\beta}^{(t+1)}
&=\left(\sum_{i=1}^mX_i^TX_i\right)^{-1}\left(\sum_{i=1}^mX_i^T(y_i-Z_i\mathbf{\mu}_1^{i})\right)\\
\widehat{\Sigma}^{(t+1)}&=\frac{1}{m}\left(\sum_{i=1}^m\mu_2^i\right)
\end{align}

This is by [216 notes](https://ucla-biostat216-2019fall.github.io/slides/16-matrixcalc/16-matrixcalc.html) for deriving derivatives.

## Q3. (20 pts) Objective of a single datum

We modify the code from HW6 to evaluate the objective, the conditional mean of $\boldsymbol{\gamma}$, and the conditional variance of $\boldsymbol{\gamma}$. Start-up code is provided below. You do _not_ have to use this code.

In [4]:
# define a type that holds an LMM datum
struct LmmObs{T <: AbstractFloat}
    # data
    y          :: Vector{T}
    X          :: Matrix{T}
    Z          :: Matrix{T}
    # posterior mean and variance of random effects γ
    μγ         :: Vector{T} # posterior mean of random effects
    νγ         :: Matrix{T} # posterior variance of random effects
    # TODO: add whatever intermediate arrays you may want to pre-allocate
    yty        :: T
    rtr        :: Vector{T}
    xty        :: Vector{T}
    zty        :: Vector{T}
    ztr        :: Vector{T}
    ztzu       :: Vector{T}
    ltztr      :: Vector{T}
    xtr        :: Vector{T}
    storage_p  :: Vector{T}
    storage_q  :: Vector{T}
    xtx        :: Matrix{T}
    ztx        :: Matrix{T}
    ztz        :: Matrix{T}
    ltztzl     :: Matrix{T}
    storage_qq :: Matrix{T}
    storage_qq2:: Matrix{T}
    storage_qq3:: Matrix{T}
    storage_I  :: Matrix{T}
end

"""
    LmmObs(y::Vector, X::Matrix, Z::Matrix)

Create an LMM datum of type `LmmObs`.
"""
function LmmObs(
    y::Vector{T}, 
    X::Matrix{T}, 
    Z::Matrix{T}) where T <: AbstractFloat
    n, p, q = size(X, 1), size(X, 2), size(Z, 2)
    μγ         = Vector{T}(undef, q)
    νγ         = Matrix{T}(undef, q, q)
    yty        = abs2(norm(y))
    rtr        = Vector{T}(undef, 1)
    xty        = transpose(X) * y
    zty        = transpose(Z) * y
    ztr        = similar(zty)
    ztzu       = Vector{T}(undef, q)
    ltztr      = similar(zty)
    xtr        = Vector{T}(undef, p)
    storage_p  = similar(xtr)
    storage_q  = Vector{T}(undef, q)
    xtx        = transpose(X) * X
    ztx        = transpose(Z) * X
    ztz        = transpose(Z) * Z
    ltztzl     = similar(ztz)
    storage_qq = similar(ztz)
    storage_qq2= similar(ztz)
    storage_qq3= similar(ztz)
    storage_I  = 1.0 * Matrix(I, q, q)
    LmmObs(y, X, Z, μγ, νγ, 
        yty, rtr, xty, zty, ztr, ztzu, ltztr, xtr,
        storage_p, storage_q, 
        xtx, ztx, ztz, ltztzl, storage_qq, storage_qq2, storage_qq3, storage_I)
end

"""
    logl!(obs::LmmObs, β, Σ, L, σ², updater = false)

Evaluate the log-likelihood of a single LMM datum at parameter values `β`, `Σ`, 
and `σ²`. The lower triangular Cholesky factor `L` of `Σ` must be supplied too.
The fields `obs.μγ` and `obs.νγ` are overwritten by the posterior mean and 
posterior variance of random effects. If `updater==true`, fields `obs.ztr`, 
`obs.xtr`, and `obs.rtr` are updated according to input parameter values. 
Otherwise, it assumes these three fields are pre-computed. 
"""
function logl!(
        obs     :: LmmObs{T}, 
        β       :: Vector{T}, 
        Σ       :: Matrix{T},
        L       :: Matrix{T},
        σ²      :: T,
        updater :: Bool = false
        ) where T <: AbstractFloat
    n, p, q = size(obs.X, 1), size(obs.X, 2), size(obs.Z, 2)
    σ²inv   = inv(σ²)
    ####################
    # Evaluate objective
    ####################
    # form the q-by-q matrix: Lt Zt Z L
    copy!(obs.ltztzl, obs.ztz)
    BLAS.trmm!('L', 'L', 'T', 'N', T(1), L, obs.ltztzl) # O(q^3)
    BLAS.trmm!('R', 'L', 'N', 'N', T(1), L, obs.ltztzl) # O(q^3)        
    # form the q-by-q matrix: M = σ² I + Lt Zt Z L
    copy!(obs.storage_qq, obs.ltztzl)
    @inbounds for j in 1:q
        obs.storage_qq[j, j] += σ²
    end
    LAPACK.potrf!('U', obs.storage_qq) # O(q^3)
    # Zt * res
    updater && BLAS.gemv!('N', T(-1), obs.ztx, β, T(1), copy!(obs.ztr, obs.zty)) # O(pq)
    # Lt * (Zt * res)
    BLAS.trmv!('L', 'T', 'N', L, copy!(obs.ltztr, obs.ztr))    # O(q^2)
    # storage_q = (Mchol.U') \ (Lt * (Zt * res))
    BLAS.trsv!('U', 'T', 'N', obs.storage_qq, copy!(obs.storage_q, obs.ltztr)) # O(q^3)
    # Xt * res = Xt * y - Xt * X * β
    updater && BLAS.gemv!('N', T(-1), obs.xtx, β, T(1), copy!(obs.xtr, obs.xty))
    # l2 norm of residual vector
    updater && (obs.rtr[1] = obs.yty - dot(obs.xty, β) - dot(obs.xtr, β))
    # assemble pieces
    logl::T = n * log(2π) + (n - q) * log(σ²) # constant term
    @inbounds for j in 1:q # log det term
        logl += 2log(obs.storage_qq[j, j])
    end
    qf    = abs2(norm(obs.storage_q)) # quadratic form term
    logl += (obs.rtr[1] - qf) * σ²inv 
    logl /= -2
    ######################################
    # TODO: Evaluate posterior mean and variance
    ######################################    
    # TODO
    
    ## νγ = σ² L(inv(M)inv(Mt))Lt
    #obs.storage_qq2 .= L
    #BLAS.trsm!('R', 'U', 'N', 'N', T(1.0), obs.storage_qq, obs.storage_qq2)
    #mul!(obs.storage_I, obs.storage_qq2, transpose(obs.storage_qq2))
    #axpby!(σ², obs.storage_I, 0.0, obs.νγ)
    BLAS.trsm!('L', 'L', 'N', 'N', 1.0, L, copy!(obs.νγ, obs.storage_I))
    BLAS.trsm!('L', 'L', 'T', 'N', 1.0, L, obs.νγ)
    axpby!(σ²inv, obs.ztz, 1.0, obs.νγ)
    ### Cholesky of (inv(Σ) + ztz/σ²)
    LAPACK.potrf!('L', obs.νγ)
    ### Inverse of (inv(Σ) + ztz/σ²)
    LAPACK.potri!('L', obs.νγ)
    LinearAlgebra.copytri!(obs.νγ, 'L')
    
    ## inv(σ²) (νγ ztr)
    mul!(obs.μγ, obs.νγ, obs.ztr, σ²inv, 0.0)
    
    ###################
    # Return
    ###################        
    return logl
end

logl!

It is a good idea to test correctness and efficiency of the single datum objective/posterior mean/var evaluator here. It's the same test datum as in HW3 and HW6.

In [5]:
Random.seed!(257)

# dimension
n, p, q = 2000, 5, 3
# predictors
X = [ones(n) randn(n, p - 1)]
Z = [ones(n) randn(n, q - 1)]
# parameter values
β  = [2.0; -1.0; rand(p - 2)]
σ² = 1.5
Σ  = fill(0.1, q, q) + 0.9I # compound symmetry 
L  = Matrix(cholesky(Symmetric(Σ)).L)
# generate y
y  = X * β + Z * rand(MvNormal(Σ)) + sqrt(σ²) * randn(n)

# form the LmmObs object
obs = LmmObs(y, X, Z);

### Correctness

In [6]:
@show logl = logl!(obs, β, Σ, L, σ², true)
@show obs.μγ
@show obs.νγ;

logl = logl!(obs, β, Σ, L, σ², true) = -3256.1793358058376
obs.μγ = [0.10608689301332777, -0.2510419060257251, -1.4653979409850368]
obs.νγ = [0.0007494356395786613 -1.2183420464422369e-6 -2.1767836829417464e-6; -1.2183420464422369e-6 0.0007542331466357765 2.1553464612468773e-5; -2.1767836829417464e-6 2.1553464612468773e-5 0.0007465271344917232]


You will lose all 20 points if following statement throws `AssertionError`.

In [7]:
@assert abs(logl - (-3256.1793358058258)) < 1e-4
@assert norm(obs.μγ - [0.10608689301333621, 
        -0.25104190602577225, -1.4653979409855415]) < 1e-4
@assert norm(obs.νγ - [
        0.0007494356395909563 -1.2183420093769967e-6 -2.176783643112221e-6; 
        -1.2183420282298223e-6 0.0007542331467601107 2.1553464632686345e-5; 
        -2.1767836636008638e-6 2.1553464641863096e-5 0.0007465271342535443
        ]) < 1e-4

### Efficiency

Benchmark for efficiency.

In [8]:
bm_obj = @benchmark logl!($obs, $β, $Σ, $L, $σ², true)

BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.021 μs …  4.367 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.067 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.071 μs ± 56.313 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                 ▁ ▅  ▆ ▇  █ █  █ ▆  ▅ ▄  ▂                   
  ▂▁▂▁▁▃▁▃▁▁▅▁▇▁▁█▁█▁▁█▁█▁▁█▁█▁▁█▁█▁▁█▁█▁▁█▁█▁▁▇▁▆▁▁▅▁▄▁▁▃▁▂ ▄
  1.02 μs        Histogram: frequency by time        1.12 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

My median runt time is 800ns. You will get full credit if the median run time is within 10μs. The points you will get are

In [9]:
clamp(10 / (median(bm_obj).time / 1e3) * 10, 0, 10)

10.0

In [10]:
# # check for type stability
@code_warntype logl!(obs, β, Σ, L, σ²)

MethodInstance for logl!(::LmmObs{Float64}, ::Vector{Float64}, ::Matrix{Float64}, ::Matrix{Float64}, ::Float64)
  from logl!(obs::LmmObs{T}, β::Vector{T}, Σ::Matrix{T}, L::Matrix{T}, σ²::T) where T<:AbstractFloat in Main at In[4]:77
Static Parameters
  T = Float64
Arguments
  #self#::Core.Const(logl!)
  obs::LmmObs{Float64}
  β::Vector{Float64}
  Σ::Matrix{Float64}
  L::Matrix{Float64}
  σ²::Float64
Body::Float64
1 ─ %1 = (#self#)(obs, β, Σ, L, σ², false)::Float64
└──      return %1



In [11]:
# using Profile

# Profile.clear()
# @profile for i in 1:10000; logl!(obs, β, Σ, L, σ²); end
# Profile.print(format=:flat)

## Q4. LmmModel type

We modify the `LmmModel` type in HW6 to hold all data points, model parameters, and intermediate arrays.

In [12]:
# define a type that holds LMM model (data + parameters)
struct LmmModel{T <: AbstractFloat}
    # data
    data :: Vector{LmmObs{T}}
    # parameters
    β      :: Vector{T}
    Σ      :: Matrix{T}
    L      :: Matrix{T}
    σ²     :: Vector{T}    
    # TODO: add whatever intermediate arrays you may want to pre-allocate
    xty    :: Vector{T}
    xtzu   :: Vector{T}
    ztu    :: Vector{T}
    xtr    :: Vector{T}
    ztr2   :: Vector{T}
    xtxinv :: Matrix{T}
    ztz2   :: Matrix{T}
end

"""
    LmmModel(data::Vector{LmmObs})

Create an LMM model that contains data and parameters.
"""
function LmmModel(obsvec::Vector{LmmObs{T}}) where T <: AbstractFloat
    # dims
    n      = size(obsvec[1].X, 1)
    p      = size(obsvec[1].X, 2)
    q      = size(obsvec[1].Z, 2)
    # parameters
    β      = Vector{T}(undef, p)
    Σ      = Matrix{T}(undef, q, q)
    L      = Matrix{T}(undef, q, q)
    σ²     = Vector{T}(undef, 1)    
    # intermediate arrays
    xty    = zeros(T, p)
    xtzu   = similar(xty)
    ztu    = Vector{T}(undef, n)
    xtr    = similar(xty)
    ztr2   = Vector{T}(undef, abs2(q))
    xtxinv = zeros(T, p, p)
    # pre-calculate \sum_i Xi^T Xi and \sum_i Xi^T y_i
    @inbounds for i in eachindex(obsvec)
        obs = obsvec[i]
        BLAS.axpy!(T(1), obs.xtx, xtxinv)
        BLAS.axpy!(T(1), obs.xty, xty)
    end
    # invert X'X
    LAPACK.potrf!('U', xtxinv)
    LAPACK.potri!('U', xtxinv)
    LinearAlgebra.copytri!(xtxinv, 'U')
    ztz2   = Matrix{T}(undef, abs2(q), abs2(q))
    LmmModel(obsvec, β, Σ, L, σ², xty, xtzu, ztu, xtr, ztr2, xtxinv, ztz2)
end

LmmModel

## Q5. Implement EM update

Let's write the key function `update_em!` that performs one iteration of EM update.

### Supplementary materials for Q5

The EM algorithm is

\begin{align}
\widehat{\sigma^2}^{(t+1)}&=\frac{1}{n}\sum_{i=1}^m\left[\lVert\widehat{w}_i\lVert_2^2+\text{Tr}(\mu_2^iZ_i^TZ_i)-2w_i^TZ_i\mu_1^i\right]\\
\widehat{\beta}^{(t+1)}
&=\left(\sum_{i=1}^mX_i^TX_i\right)^{-1}\mathbf{X}^T\left(\mathbf{y}-\mathbf{Z}\mathbf{\mu}_1\right)\\
&=\left(\sum_{i=1}^mX_i^TX_i\right)^{-1}\left(\sum_{i=1}^mX_i^T(y_i-Z_i\mathbf{\mu}_1^{i})\right)\\
\widehat{\Sigma}^{(t+1)}&=\frac{1}{m}\left(\sum_{i=1}^m\mu_2^i\right)
\end{align}

where

\begin{align}
\mu_{1}^{i}=\mathbb{E}\left(\gamma_i\mid \boldsymbol{\beta}^{(t)}, \boldsymbol{\Sigma}^{(t)}, \sigma^{2(t)}\right)&=\sigma^{-2(t)}\left(\frac{Z_i^TZ_i}{\sigma^{2(t)}}+(\boldsymbol{\Sigma}^{(t)})^{-1}\right)^{-1}Z_i^Tw_i\\
\mu_{2}^{i}=\mathbb{E}\left(\gamma_i\gamma_i^T\mid \boldsymbol{\beta}^{(t)}, \boldsymbol{\Sigma}^{(t)}, \sigma^{2(t)}\right)&=\left(\frac{Z_i^TZ_i}{\sigma^{2(t)}}+(\boldsymbol{\Sigma}^{(t)})^{-1}\right)^{-1}+\sigma^{-4(t)}\left(\frac{Z_i^TZ_i}{\sigma^{2(t)}}+(\boldsymbol{\Sigma}^{(t)})^{-1}\right)^{-1}Z_i^Tw_iw_i^TZ_i\left(\frac{Z_i^TZ_i}{\sigma^{2(t)}}+(\boldsymbol{\Sigma}^{(t)})^{-1}\right)^{-1}
\end{align}

In [13]:
"""
    update_em!(m::LmmModel, updater::Bool = false)

Perform one iteration of EM update. It returns the log-likelihood calculated 
from input `m.β`, `m.Σ`, `m.L`, and `m.σ²`. These fields are then overwritten 
by the next EM iterate. The fields `m.data[i].xtr`, `m.data[i].ztr`, and 
`m.data[i].rtr` are updated according to the resultant `m.β`. If `updater==true`, 
the function first updates `m.data[i].xtr`, `m.data[i].ztr`, and 
`m.data[i].rtr` according to `m.β`. If `updater==false`, it assumes these fields 
are pre-computed.
"""
function update_em!(m::LmmModel{T}, updater::Bool = false) where T <: AbstractFloat
    logl = zero(T)
    M = length(m.data)
    # TODO: update m.β
    fill!(m.xtzu, 0.0)
    @inbounds for i in 1:length(m.data)
        obs = m.data[i]
        logl += logl!(obs, m.β, m.Σ, m.L, m.σ²[1], true)
        
        mul!(obs.storage_p, transpose(obs.ztx), obs.μγ)
        BLAS.axpby!(T(1), obs.storage_p, T(1), m.xtzu)
    end
    axpby!(1.0, m.xty, -1.0, m.xtzu)
    BLAS.gemv!('N', 1.0, m.xtxinv, m.xtzu, 0.0, m.β)
    
    # TODO: update m.data[i].ztr, m.data[i].xtr, m.data[i].rtr
    sum_n = 0.0
    fill!(m.σ², 0.0)
    fill!(m.Σ, 0.0)
    
    @inbounds for i in 1:M
        obs = m.data[i]
        
        if updater
        # update m.data[i].ztr, m.data[i].xtr, m.data[i].rtr
        # Zt * res
            BLAS.gemv!('N', T(-1), obs.ztx, m.β, T(1), copy!(obs.ztr, obs.zty))
        # Xt * res = Xt * y - Xt * X * β
            BLAS.gemv!('N', T(-1), obs.xtx, m.β, T(1), copy!(obs.xtr, obs.xty))
        # l2 norm of residual vector
            obs.rtr[1] = obs.yty - dot(obs.xty, m.β) - dot(obs.xtr, m.β)
        end
        
        # update m.σ²
        sum_n += size(obs.y)[1]
        m.σ²[1] += obs.rtr[1]
        m.σ²[1] -= 2 * dot(obs.zty, obs.μγ)
        m.σ²[1] += 2 * dot(m.β, obs.storage_p)
        mul!(obs.storage_qq3, obs.νγ, obs.ztz)
        m.σ²[1] += tr(obs.storage_qq3)
        BLAS.gemv!('N', T(1), obs.ztz, obs.μγ, T(0), obs.storage_q)
        m.σ²[1] += dot(obs.μγ, obs.storage_q)
        
        # update m.Σ 
        axpby!(1.0, obs.νγ, 1.0, m.Σ)
        BLAS.ger!(1.0, obs.μγ, obs.μγ, m.Σ)
        
    end
    
    # TODO: update m.σ²
    m.σ²[1] /= sum_n
    
    # update m.L
    lmul!(1/M, m.Σ)
    LAPACK.potrf!('L', copy!(m.L, m.Σ))
    
    # return log-likelihood at input parameter values
    logl
end

update_em!

## Q6. (30 pts) Test data

Let's generate a synthetic longitudinal data set (same as HW6) to test our algorithm.

In [14]:
Random.seed!(257)

# dimension
m      = 1000 # number of individuals
ns     = rand(1500:2000, m) # numbers of observations per individual
p      = 5 # number of fixed effects, including intercept
q      = 3 # number of random effects, including intercept
obsvec = Vector{LmmObs{Float64}}(undef, m)
# true parameter values
βtrue  = [0.1; 6.5; -3.5; 1.0; 5]
σ²true = 1.5
σtrue  = sqrt(σ²true)
Σtrue  = Matrix(Diagonal([2.0; 1.2; 1.0]))
Ltrue  = Matrix(cholesky(Symmetric(Σtrue)).L)
# generate data
for i in 1:m
    # first column intercept, remaining entries iid std normal
    X = Matrix{Float64}(undef, ns[i], p)
    X[:, 1] .= 1
    @views Distributions.rand!(Normal(), X[:, 2:p])
    # first column intercept, remaining entries iid std normal
    Z = Matrix{Float64}(undef, ns[i], q)
    Z[:, 1] .= 1
    @views Distributions.rand!(Normal(), Z[:, 2:q])
    # generate y
    y = X * βtrue .+ Z * (Ltrue * randn(q)) .+ σtrue * randn(ns[i])
    # form a LmmObs instance
    obsvec[i] = LmmObs(y, X, Z)
end
# form a LmmModel instance
lmm = LmmModel(obsvec);

### Correctness

Evaluate log-likelihood and gradient at the true parameter values.

In [15]:
copy!(lmm.β, βtrue)
copy!(lmm.Σ, Σtrue)
copy!(lmm.L, Ltrue)
lmm.σ²[1] = σ²true
@show obj1 = update_em!(lmm, true)
@show lmm.β
@show lmm.Σ
@show lmm.L
@show lmm.σ²
println()
@show obj2 = update_em!(lmm, false)
@show lmm.β
@show lmm.Σ
@show lmm.L
@show lmm.σ²

obj1 = update_em!(lmm, true) = -2.8400684383699736e6
lmm.β = [0.10003613673625007, 6.500382871080181, -3.499864634211224, 0.9997124657606651, 4.999230851463544]
lmm.Σ = [1.9903882760455252 0.06862095707037451 0.05347290179472647; 0.06862095707037451 1.2813220461216897 -0.0904491332490633; 0.05347290179472647 -0.0904491332490633 0.9435400745724168]
lmm.L = [1.4108112120498353 0.06862095707037451 0.05347290179472647; 0.048639361868036066 1.130909482937852 -0.0904491332490633; 0.037902237619045515 -0.08160924927472396 0.9671832429217858]
lmm.σ² = [1.4987367279243728]

obj2 = update_em!(lmm, false) = -2.8400604605420595e6
lmm.β = [0.10007136571174416, 6.500383550631511, -3.4998642980415435, 0.9997119269497639, 4.999229480978786]
lmm.Σ = [1.9903775380783657 0.06870107683947124 0.05354351750443938; 0.06870107683947124 1.2814409213711735 -0.09059223997463764; 0.05354351750443938 -0.09059223997463764 0.9434431688085724]
lmm.L = [1.410807406444397 0.06870107683947124 0.05354351750443938; 0.0486

1-element Vector{Float64}:
 1.4987403307694511

Test correctness. You will loss all 30 points if following code throws `AssertError`.

In [16]:
@assert abs(obj1 - (-2.840068438369969e6)) < 1e-4
@assert abs(obj2 - (-2.84006046054206e6)) < 1e-4

### Efficiency

Test efficiency of EM update.

In [17]:
bm_emupdate = @benchmark update_em!($lmm, true) setup=(
    copy!(lmm.β, βtrue);
    copy!(lmm.Σ, Σtrue);
    copy!(lmm.L, Ltrue);
    lmm.σ²[1] = σ²true)

BenchmarkTools.Trial: 3688 samples with 1 evaluation.
 Range (min … max):  1.318 ms …  1.518 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.353 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.355 ms ± 19.785 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

         ▁▁▂▁▁   ▁▄▅▇█▇▅▆▂                                    
  ▁▁▁▃▄▅██████▆▇▇██████████▆▅▄▃▄▃▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▁▁▁▁▁▁▁▂▁▁▁▁ ▃
  1.32 ms        Histogram: frequency by time        1.43 ms <

 Memory estimate: 0 bytes, allocs estimate: 0.

My median run time is 1ms. You will get full credit if your median run time is within 10ms. The points you will get are

In [18]:
clamp(10 / (median(bm_emupdate).time / 1e6) * 10, 0, 10)

10.0

### Memory

You will lose 1 point for each 100 bytes memory allocation. So the points you will get is

In [19]:
clamp(10 - median(bm_emupdate).memory / 100, 0, 10)

10.0

## Q7. Starting point

We use the same least squares estimates as in HW6 as starting point. 

In [20]:
"""
    init_ls!(m::LmmModel)

Initialize parameters of a `LmmModel` object from the least squares estimate. 
`m.β`, `m.L`, and `m.σ²` are overwritten with the least squares estimates.
"""
function init_ls!(m::LmmModel{T}) where T <: AbstractFloat
    p, q = size(m.data[1].X, 2), size(m.data[1].Z, 2)
    # LS estimate for β
    mul!(m.β, m.xtxinv, m.xty)
    # LS etimate for σ2 and Σ
    rss, ntotal = zero(T), 0
    fill!(m.ztz2, 0)
    fill!(m.ztr2, 0)    
    @inbounds for i in eachindex(m.data)
        obs = m.data[i]
        ntotal += length(obs.y)
        # update Xt * res
        BLAS.gemv!('N', T(-1), obs.xtx, m.β, T(1), copy!(obs.xtr, obs.xty))
        # rss of i-th individual
        rss += obs.yty - dot(obs.xty, m.β) - dot(obs.xtr, m.β)
        # update Zi' * res
        BLAS.gemv!('N', T(-1), obs.ztx, m.β, T(1), copy!(obs.ztr, obs.zty))
        # Zi'Zi ⊗ Zi'Zi
        kron_axpy!(obs.ztz, obs.ztz, m.ztz2)
        # Zi'res ⊗ Zi'res
        kron_axpy!(obs.ztr, obs.ztr, m.ztr2)
    end
    m.σ²[1] = rss / ntotal
    # LS estimate for Σ = LLt
    LAPACK.potrf!('U', m.ztz2)
    BLAS.trsv!('U', 'T', 'N', m.ztz2, m.ztr2)
    BLAS.trsv!('U', 'N', 'N', m.ztz2, m.ztr2)
    copyto!(m.Σ, m.ztr2)
    copy!(m.L, m.Σ)
    LAPACK.potrf!('L', m.L)
    for j in 2:q, i in 1:j-1
        m.L[i, j] = 0
    end
    m
end

"""
    kron_axpy!(A, X, Y)

Overwrite `Y` with `A ⊗ X + Y`. Same as `Y += kron(A, X)` but
more memory efficient.
"""
function kron_axpy!(
        A::AbstractVecOrMat{T},
        X::AbstractVecOrMat{T},
        Y::AbstractVecOrMat{T}
        ) where T <: Real
    m, n = size(A, 1), size(A, 2)
    p, q = size(X, 1), size(X, 2)
    @assert size(Y, 1) == m * p
    @assert size(Y, 2) == n * q
    @inbounds for j in 1:n
        coffset = (j - 1) * q
        for i in 1:m
            a = A[i, j]
            roffset = (i - 1) * p            
            for l in 1:q
                r = roffset + 1
                c = coffset + l
                for k in 1:p                
                    Y[r, c] += a * X[k, l]
                    r += 1
                end
            end
        end
    end
    Y
end

kron_axpy!

In [21]:
init_ls!(lmm)
@show lmm.β
@show lmm.Σ
@show lmm.L
@show lmm.σ²

lmm.β = [0.18207934611476326, 6.50048070099372, -3.4979107842091595, 1.0011132962297953, 5.0002519857919285]
lmm.Σ = [1.9794302836685058 0.07258461003916691 0.05717147035274018; 0.07258461003916693 1.2840385734767714 -0.07707942768978565; 0.05717147035274018 -0.07707942768978565 0.9509885905046899]
lmm.L = [1.4069222734993236 0.0 0.0; 0.05159105901325531 1.1319792118703693 0.0; 0.04063584138912111 -0.06994463586493146 0.9718256360134827]
lmm.σ² = [5.709004733413668]


1-element Vector{Float64}:
 5.709004733413668

## Q8. Estimation by EM

We write a function `fit!` that implements the EM algorithm for estimating LMM.

In [22]:
"""
    fit!(m::LmmModel)

Fit an `LmmModel` object by MLE using a EM algorithm. Start point 
should be provided in `m.β`, `m.σ²`, `m.L`.
"""
function fit!(
        m       :: LmmModel;
        maxiter :: Integer       = 10_000,
        ftolrel :: AbstractFloat = 1e-12,
        prtfreq :: Integer       = 0
    )
    obj = update_em!(m, true)
    for iter in 0:maxiter
        obj_old = obj
        # EM update
        obj = update_em!(m, false)
        # print obj
        prtfreq > 0 && rem(iter, prtfreq) == 0 && println("iter=$iter, obj=$obj")
        # check monotonicity
        obj < obj_old && (@warn "monotonicity violated")
        # check convergence criterion
        (obj - obj_old) < ftolrel * (abs(obj_old) + 1) && break
        # warning about non-convergence
        iter == maxiter && (@warn "maximum iterations reached")
    end
    m
end


fit!

## Q9. (20 pts) Test drive

Now we can run our EM algorithm to compute the MLE.

In [23]:
# initialize from least squares
init_ls!(lmm)

@time fit!(lmm, prtfreq = 1);

println("objective value at solution: ", update_em!(lmm)); println()
println("solution values:")
@show lmm.β
@show lmm.σ²
@show lmm.L * transpose(lmm.L)

iter=0, obj=-2.840068882178379e6
iter=1, obj=-2.8400587867554654e6
iter=2, obj=-2.840058786725546e6
iter=3, obj=-2.8400587867254787e6
  0.028271 seconds (57.70 k allocations: 3.161 MiB, 71.43% compilation time)
objective value at solution: -2.840058786725409e6

solution values:
lmm.β = [0.18207855803149936, 6.500383547392576, -3.499864296224195, 0.9997119252318636, 4.999229481214088]
lmm.σ² = [1.4987345519183402]
lmm.L * transpose(lmm.L) = [1.9910121170848079 0.13511176181337323 0.10338307121955914; 0.13511176181337323 1.2896479966475474 -0.1782018729425176; 0.10338307121955914 -0.1782018729425176 0.94344304716562]


3×3 Matrix{Float64}:
 1.99101    0.135112   0.103383
 0.135112   1.28965   -0.178202
 0.103383  -0.178202   0.943443

### Correctness

You get 10 points if the following code does not throw `AssertError`.

In [24]:
# objective at solution should be close enough to the optimal
@assert update_em!(lmm) > -2.840059e6

### Efficiency

My median run time 5ms. You get 10 points if your median run time is within 1s.

In [25]:
bm_em = @benchmark fit!($lmm) setup = (init_ls!(lmm))

BenchmarkTools.Trial: 755 samples with 1 evaluation.
 Range (min … max):  6.216 ms …   7.007 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     6.324 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.364 ms ± 121.267 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

      ▁ ▃▆█▇▅▂ ▁                                               
  ▃▄▇▅██████████▇███▆▅▅▃▄▃▄▃▃▃▂▃▃▃▂▃▃▃▃▂▂▃▂▃▃▃▃▃▂▁▂▃▂▃▁▃▂▂▁▂▂ ▄
  6.22 ms         Histogram: frequency by time        6.81 ms <

 Memory estimate: 0 bytes, allocs estimate: 0.

In [26]:
# this is the points you get
clamp(1 / (median(bm_em).time / 1e9) * 10, 0, 10)

10.0

## Q10. (10 pts) EM vs Newton type algorithms

Contrast EM algorithm to the Newton type algorithms (gradient free, gradient based, using Hessian) in HW6, in terms of the stability, convergence rate (how fast the algorithm is converging),  final objective value, total run time, derivation, and implementation efforts. 

## Sol Q10

| Items | EM algorithm | Newton-type algorithms |
|---|---|---|
| stability | Monotonicity (always increasing) |Not guaranteed to be increasing|
| Convergence rate | Linear (associated with the eigenvalues of Fisher information matrix) | Fast (e.g., second order for Newton-Raphson) |
| Final objective value | -2.840058e6 | -2.83426e6 |
| Total run time | 7.007 ms | 0.61 ms |
| Derivation| CRAZY and PAINFUL!| Matured technology (according to Dr. Zhou and Dr. Vandenberghe)|
|Implementation efforts| Easy (just a few lines)| Extremelly *PAINFUL* to get gradients and Hessian|